### PR TITLE
Closes #4923: Deny permissions when the dialog is dismissed

### LINKED ISSUE #4923: Users can dismiss site permissions without resolving the Promise to the site
STR:

- Open https://permission.site
- Request Notification permission
- Dismiss the doorhanger by clicking anywhere else

Expected result:
- The "Notifications" button should turn red because the Promise is rejected

Actual result:
- Nothing happens, the Promise is left dangling and never resolved because the user won't be able to go back to the prompt

This is a web compat and dev UX issue, because the website can not possibly tell whether the user is still making up their mind or has already dismissed the permission doorhanger. This way many users will be left on interstitial dialogs along the lines of "please accept the permission prompt before continuing". We had this problem on desktop a while ago.

On desktop we solved this by making the doorhangers un-dismissable, which is somewhat of a user annoyance, but users would otherwise dismiss the doorhanger too easily, leading to confusion about where it's gone.

On mobile, I think the better solution/UX would be to reject the promise when the dialog is dismissed. I think that the prominent way the doorhanger is shown on mobile reduces the likelihood that a user would accidentally dismiss the prompt and it seems like we should allow users to escape the prompt by tapping anywhere.

### PR DESCRIPTION
**Pull request by johannhof**
---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)

**I think this passes, but ktlint is throwing weird errors about missing newlines in my console**

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one

**Does this need a changelog entry?**

- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/feature/sitepermissions/src/test/java/mozilla/components/feature/sitepermissions/SitePermissionsDialogFragmentTest.kt [hunk 1]
```diff
@@ -148,6 +148,36 @@ class SitePermissionsDialogFragmentTest {
...
+        doReturn(mockFragmentManager()).`when`(fragment).fragmentManager
+
+        val dialog = fragment.onCreateDialog(null)
+        dialog.show()
+        // XXX(johannh): Calling this method makes the test a bit pointless,
+        // would be nicer to actually dismiss the dialog, but that
+        // seems all mocked away.
+        fragment.onDismiss(dialog)
+
+        verify(mockFeature).onDismiss("sessionId")
```
Comment (Amejia481) [inline components/feature/sitepermissions/src/test/java/mozilla/components/feature/sitepermissions/SitePermissionsDialogFragmentTest.kt]: Nice monologue :)

You could remove the dialog creation and call
`fragment.onDismiss(mock())`


The test could end up looking like this:

```kotlin
    @Test
    fun `dismissing the dialog notifies the feature`() {
        val mockFeature: SitePermissionsFeature = mock()

        val fragment = spy(
                SitePermissionsDialogFragment.newInstance(
                        "sessionId",
                        "title",
                        R.drawable.notification_icon_background,
                        mockFeature,
                        shouldShowDoNotAskAgainCheckBox = false,
                        shouldSelectDoNotAskAgainCheckBox = false
                )
        )

        fragment.feature = mockFeature

        doReturn(testContext).`when`(fragment).requireContext()
        doReturn(mockFragmentManager()).`when`(fragment).fragmentManager

        fragment.onDismiss(mock())

        verify(mockFeature).onDismiss("sessionId")
    }
```

**Comment by NotWoods** [comment]
bors r=Amejia481

**Comment by Amejia481** [comment]
I'm not sure if @johannhof  wants to update this section https://github.com/mozilla-mobile/android-components/pull/4924#discussion_r341710446
bors r-

**Comment by NotWoods** [comment]
Whoops, didn't see the comment. Sorry!

**Comment by Amejia481** [comment]
No worries :)

**Comment by johannhof** [comment]
bors r=Amejia481

**Comment by johannhof** [comment]
¯\_(ツ)_/¯

**Comment by NotWoods** [comment]
bors r=Amejia481

**Comment by NotWoods** [comment]
bors retry

**Pull request merged**